File: csci41/lec29/Counter.java
// there are no method/function declarations in Java--you just start
// implementing!
public class Counter extends Object { // every class in java by
                                      // default extends Object
  // stores the actual count
  private int count;

  // Java has no default parameters, so you can't do something like:
  //   public Counter(int startingValue = 0) { ... }
  // You have to make two separate constructors

  public Counter() {
    count = 0;
  }

  public Counter(int startingValue) {
    count = startingValue;
  }

  public int getCount() {
    return count;
  }

  public void inc() {
    count++;
  }

  public void reset() {
    count = 0;
  }

  public void reset(int newValue) {
    count = newValue;
  }

  // Java has no operator overloading--you have to use words.
  // There are also no const parameters--you have to do some weird
  // tricks if you want to guarantee other doesn't get modified.
  public Counter add(Counter other) {
    Counter ret = new Counter(count + other.count);
    return ret;
  }

  public static void main(String[] args) {
    Counter c = new Counter();
    c.inc();
    c.inc();
    c.inc();
    System.out.println(c.getCount());

    Counter c2 = new Counter(42);
    c2.inc();
    System.out.println(c2.getCount());

    Counter c3 = c.add(c2);
    System.out.println(c3.getCount());
  }
}

File: csci41/lec29/counter.py
class Counter:
    # member variables get defined in the constructor

    # constructors use __init__ instead of the name of the class
    # every method (including the constructor) it takes "self" as the
    # first parameter--that is the "this" from C++
    def __init__(self):
        self.count = 0 # this defines the member var

    def inc(self):
        # there's no ++ in python
        self.count += 1

    def getCount(self):
        return self.count

# there is no new in Python, you just call the constructor
# notice that self is not being used--python includes it for you!!!
c = Counter() # equivalent to Counter c = new Counter(); from Java
c.inc() # --> inc(c)
c.inc()
c.inc()
print(c.getCount())
